[ML] Write downloaded model parts async

davidkyle opened the pull request:
It has been observed that downloading and installing the built in `.elser_model_2` and `.multilingual-e5-small` models is much slower than expected. The cause is in the `ModelImporter` class which downloads the model definition in 1MB chunks then blocks as the model part is written to the index. 

The download server supports the [Range header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range), to speed up the download and install multiple connections are made to the server each asking for a separate range. A dedicated thread handle downloading and index the parts in each range. 5 connections are used in this PR, reading a 1MB chunk at a time to limit the amount of memory used.

The final part of the model definition must be written last as it causes an index refresh making the full model definition visible, if the refresh occurs before all parts are written and not all the parts are visible then deploying the model will fail. This is achieved by indexing the final part only once all the other streams have completed. 

There is a problem with calculating the SHA 256 Message Digest of the downloaded model. For one the MessageDigest is not thread safe, more problematically the model parts are not downloaded sequentially and the resulting digest changes depending on the order in which the parts are downloaded.

davidkyle committed:
Write model parts async

davidkyle committed:
Update docs/changelog/111684.yaml

elasticsearchmachine commented:
Hi @davidkyle, I've created a changelog YAML for you.

jimczi commented on x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/ModelImporter.java:
```diff
@@ -136,17 +146,19 @@ private void uploadVocabulary() throws URISyntaxException {
         }
 
-        client.execute(action, request).actionGet();
+        requestLimiter.acquire();
```
We'll still block the thread for the duration of the entire download. Since we know the number of parts, would it be possible to start N downloads concurrently and let the completion listener handles the continuations/failures? That's how we handle throttling in other transport requests. Blocking a thread to wait for the end of a network call is generally not recommended.

davidkyle replied:
It is not possible to slice the download into parts for concurrent download as the InputStream does not support that, only 1 thread can read from the InputStream at a time. I've switched to a RefCountingListener to collect the responses which avoids the block waiting for all the writes to finish but I still have a block on acquiring the semaphore when there are too many inflight requests. That is not ideal but it is a vast improvement. I'll get some opinions on how to remove the blocking wait on the semaphore.

jimczi commented on x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/ModelImporter.java:
```diff
@@ -119,11 +127,13 @@ public void doImport() throws URISyntaxException, IOException, ElasticsearchStat
-        executeRequestIfNotCancelled(PutTrainedModelDefinitionPartAction.INSTANCE, finalModelPartRequest);
+        executeRequestIfNotCancelled(PutTrainedModelDefinitionPartAction.INSTANCE, finalModelPartRequest, releasingListener);
+        logger.info("waiting for finish");
+        requestLimiter.acquire(MAX_IN_FLIGHT_REQUESTS); // cannot acquire until all inflight requests have completed
```
Same here, we shouldn't block the thread to wait for the end of all downloads. A completion listener should be used to orchestrate the remaining work when the last download completed.

davidkyle committed:
Pass a listener to import

davidkyle committed:
Ref counting WIP

davidkyle committed:
use ref counting listener

davidkyle committed:
tidying

davidkyle committed:
add tests

davidkyle commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into background-download-write

elasticsearchmachine commented:
Pinging @elastic/ml-core (Team:ML)

davidkyle committed:
Add download threadpool

davidkyle committed:
less blocking

davidkyle committed:
tidy up

jimczi commented on x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/ModelImporter.java:
```diff
@@ -35,61 +36,173 @@
+                    chunkIterator.getTotalParts(),
+                    size,
+                    definition,
+                    countingListener.acquire(r -> executorService.execute(() -> doNextPart(size, chunkIterator, countingListener)))
```
That's lock free because we rely on the thread pool having only one thread? 
Isn't it equivalent to what we have now where a single thread does the download+upload?

davidkyle replied:
> That's lock free because we rely on the thread pool having only one thread? 

Yes, it only works because the single thread executor limits concurrent read access to the input stream.

> Isn't it equivalent to what we have now where a single thread does the download+upload?

The upload (index part) is now async which is where we get the speed up and the indexing occurs on a different thread.

jimczi commented on x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/ModelImporter.java:
```diff
@@ -35,61 +36,173 @@
+                config.getModelRepository(),
+                config.getPackagedModelId() + ModelLoaderUtils.MODEL_FILE_EXTENSION
+            );
+            modelInputStream = ModelLoaderUtils.getInputStreamFromModelRepository(uri);
```
Looking at the upload server, it accepts range queries with the form `Range: bytes=0-1024` so we should be able to download concurrently (or read files concurrently using a random reader)? If downloading async is too complicated we can use the new download threadpool. That would also limit the number of download we have. Regarding the current blocking part, it happens on the ml utility thread pool, that's what worries me the most. I think we need to move the blocking part to the download thread pool

jimczi replied:
With the range download we can have multiple input streams doing the download in parallel, each one doing a single range. Currently we limit the download to one input stream because we retrieve the entire file from one connection. With the range header we can use one connection for each part (limited to N connections in parallel with a dedicated thread pool if we want). 
I tried with the following:
```
curl --location 'https://ml-models.elastic.co/multilingual-e5-small.pt' \
--header 'Range: bytes=0-1024'
```
That returns the first 1024 bytes so we can split into multiple tasks, each responsible to download and store a range of the file instead of splitting from a single input stream?

davidkyle committed:
more tests

davidkyle commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into background-download-write

davidkyle committed:
remove unused

davidkyle committed:
fix the tests

davidkyle committed:
5 in flight requests

davidkyle commented:
In classic cloud this change has taken the model download & install time down from 30 seconds to [7 - 10] seconds with the total time to download and deploy ELSER optimised at 14 seconds. 

In serverless the download & install time is down to 21 seconds and the total time to download and deploy ELSER optimised 31 seconds. 

Those severless numbers aren't good enough, I will try another approach

davidkyle committed:
use another threadpool for writes

davidkyle commented:
Using a dedicated executor for writing the model parts *and* a separate executor for reading the model stream did not improve the download & install time. In ESS the time to download and deploy `.elser_model_2_linux-x86_64` averaged at 14 seconds but on serverless the same process stubbornly remains at over 30 seconds (38s - 41s). Next up is to try the range header.

davidkyle committed:
Revert "use another threadpool for writes"

This reverts commit 70ffe07c60138543edef37f245bf5be33758a478.

davidkyle committed:
use range request

davidkyle committed:
Merge branch 'main' into background-download-write

davidkyle committed:
Use multiple connections

davidkyle commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into background-download-write

davidkyle committed:
Tidy comments

davidkyle committed:
short threads

davidkyle commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into background-download-write

kingherc committed:
Enable markSupported in SlicedInputStream (#112563)

This means openSlice must be able to be called for a previous slice
even if the previous slice has been closed.

We disable the support for current implementations to keep
their old behavior just to be sure.

Relates ES-9248

original-brownbear committed:
Make two SubReaderWrapper implementations singletons (#112596)

random find:

No state in either of these, no need to recreate them for every
directory. Saves a few cycles and makes it more obvious that there's no
state in these.

original-brownbear committed:
Two speedups to IndexNameExpressionResolver (#112486)

Two obvious speedups to the IndexNameExpressionResolver where we can
defer an expensive lookup from either the indices lookup or the thread
context to if and when we actually need it.

cbuescher committed:
Fix failing LangMustacheClientYamlTestSuiteIT yamlRestTestV7CompatTest (#112683)

The error message changed slightly to include the failure location. While this
was added to the main test yml file, the yamlRestTestV7CompatTest resources seem
to get copied from v7 files and we then modify them if we expect any changes.

Closes #112580

DaveCTurner committed:
Fix trappy timeouts in downsample action (#112734)

Relates #107984

elasticsearchmachine committed:
Mute org.elasticsearch.smoketest.DocsClientYamlTestSuiteIT test {yaml=reference/cluster/stats/line_1450} #112732

DaveCTurner committed:
Introduce repository integrity verification API (#112348)

Adds an API which scans all the metadata (and optionally the raw data)
in a snapshot repository to look for corruptions or other
inconsistencies.

Closes https://github.com/elastic/elasticsearch/issues/52622 Closes
ES-8560

benwtrent committed:
JSON parse failures should be 4xx codes (#112703)

It seemed if there wasn't any text to parse, this is not an internal
issue but instead an argument issue.

I simply changed the exception thrown. If we don't agree with this, I
can adjust `query` parsing directly, but this seemed like the better
choice.

closes: https://github.com/elastic/elasticsearch/issues/112296

DaveCTurner committed:
Handle null exception message in `TestCluster#wipe` (#112741)

Fixing noise seen in an unrelated test failure.

parkertimmins committed:
Add TaskManager to pluginServices (#112687)

Provide access to the TaskManager in plugins so that a RemovedTaskListener can be registered. This will allow a plugin to monitor when submitted tasks complete.

gmarouli committed:
Introduce data stream options and failure store configuration classes (#109515)

In order to facilitate enabling and disabling the failure store &
component template composition, we introduce new metadata classes that
can support a more extensible failure store configuration.

We would like to introduce **data stream options**. Data stream options
capture the configuration of data stream level (smaller and larger)
features, such the failure store and in the future data stream
lifecycle. They are different than settings because they are applied on
a data stream level and not per backing index.

This PR is only setting the basic classes to enable follow up PRs that
will actually use them. 

Examples, these are not final, they are only used to help visualise a
potential direction:

```
GET _data_stream/my-*/_options
{
  "data_streams": [
    {
      "name": "my-non-opinionated-ds",
      "options": {
      }
    },
    {
      "name": "my-fs",
      "options": {
        "failure_store": {
          "enabled": true
        }
      }
    },
    {
      "name": "my-no-fs",
      "options": {
        "failure_store": {
          "enabled": false
        }
      }
    }
  ]
}

// If we decide to add lifecycle here too:
PUT _data_stream/my-fs/_options
{
   "failure_store": {
     "enabled": true
   },
   "lifecycle": {
   }
}
```

What we see above are 3 data streams:

- `my-fs` with the failure store explicitly enabled
- `my-no-fs` with the failure store explicitly disabled, and
- `my-non-opinionated-ds` which does not specify what to do with the failure store, so for now it means failure store disabled but that could change in the future.

Template composition examples pending

craigtaverner committed:
Support widening of numeric types in union-types (#112610)

* Support widening of numeric types in union-types

Only two lines of this PR are the actual fix.
All the rest is updating the CSV-spec testing infrastructure to make it easier to test this, and adding the tests.
The refactoring involve some cleanup and simplifications also.
This update allows us to add alternative mappings of existing data files without copying the files and changing the header line.
Some of the existing union-types test files were deleted as a result, which is a step in the right direction.

* Update docs/changelog/112610.yaml

* Link capability to PR

maxcold committed:
add CDR related data streams to kibana_system priviliges (#112655)

mark-vieira committed:
Bump Elasticsearch version to 9.0.0 (#112570)

nik9000 committed:
ESQL: Compute support for filtering ungrouped aggs (#112717)

Adds support to the compute engine for filtering which positions are
processed by ungrouping aggs. This should allow syntax like:

```
| STATS
       success = COUNT(*) WHERE 200 <= response_code AND response_code < 300,
      redirect = COUNT(*) WHERE 300 <= response_code AND response_code < 400,
    client_err = COUNT(*) WHERE 400 <= response_code AND response_code < 500,
    server_err = COUNT(*) WHERE 500 <= response_code AND response_code < 600,
   total_count = COUNT(*)
```

We could translate the WHERE expression into an `ExpressionEvaluator`
and run it, then plug it into the filtering support added in this PR.

The actual filtering is done by creating a `FilteredAggregatorFunction`
which wraps a regular `AggregatorFunction` first executing the filter
against the incoming `Page` and then passing the resulting mask to the
`AggregatorFunction`. We've then added a `mask` to
`AggregatorFunction#process` which each aggregation function must use
for filtering.

We keep the unfiltered behavior by sending a constant block with `true`
in it. Each agg detects this and takes an "unfiltered" path, preserving
the original performance.

Importantly, when you don't turn this on it doesn't effect performance:

```
 (blockType)  (grouping)   (op)  Score    Error -> Score    Error  Units
vector_longs        none  count  0.007 ±  0.001 -> 0.007 ±  0.001  ns/op
vector_longs        none    min  0.123 ±  0.004 -> 0.128 ±  0.005  ns/op
vector_longs       longs  count  4.311 ±  0.192 -> 4.218 ±  0.053  ns/op
vector_longs       longs    min  5.476 ±  0.077 -> 5.451 ±  0.074  ns/op
```

ChrisHegarty committed:
Bump Elasticsearch to a minimum of JDK 21 (#112252)

jimczi commented on x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/MachineLearningPackageLoader.java:
```diff
@@ -81,6 +85,24 @@ public List<NamedWriteableRegistry.Entry> getNamedWriteables() {
+
+    public static FixedExecutorBuilder modelDownloadExecutor(Settings settings) {
+        // Threadpool with a fixed size of 1 thread for
+        // downloading the model definition files
```
The comment seems outdated

jimczi commented on x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/ModelLoaderUtils.java:
```diff
@@ -61,6 +66,73 @@ final class ModelLoaderUtils {
+        }
+    }
+
+    static class HttStreamChunker {
```
nit: 
```suggestion
    static class HttpStreamChunker {
```

elasticsearchmachine committed:
Mute org.elasticsearch.repositories.blobstore.testkit.integrity.RepositoryVerifyIntegrityIT testCorruption #112769

jimczi reviewed (approved):
Thanks for the all the iteration @davidkyle!
This approach looks the best option when using synchronous connections. 
We should discuss how we could switch to an asynchronous http client but that's for a follow up ;).

lcawl committed:
[DOCS] Augment installation warnings (#112756)

stefnestor committed:
(Doc+) CAT Nodes default columns (#112715)

👋 howdy, team!

1. Related to https://github.com/elastic/dev/issues/2631, highlights customers are usually seeking `heap.percent` instead of `ram.percent`
2. Aligns the claimed "(Default)" columns in doc to what returned for v8.15.1 test cluster

stefnestor committed:
(Doc+) Terminating Exit Codes (#112530)

👋 howdy, team! Mini PR to cross-replicate [this knowledge article](https://support.elastic.co/knowledge/6610ba83) about Elasticsearch's exit codes which expands [this ES doc section](https://www.elastic.co/guide/en/elasticsearch/reference/master/stopping-elasticsearch.html#fatal-errors).

mark-vieira committed:
Fix verifyVersions task (#112765)

stefnestor committed:
(Doc+) Inference Pipeline ignores Mapping Analyzers (#112522)

* (Doc+) Inference Pipeline ignores Mapping Analyzers

From internal Dev feedback (will cross-link after), this updates that inference processors within ingest pipelines run before mapping analyzers effectively ignoring them. So if users want analyzers to take effect, they would need to select the analyzer's ingest pipeline process equivalent and run it higher in flow than the inference processor.

---------

Co-authored-by: István Zoltán Szabó <istvan.szabo@elastic.co>

dnhatn committed:
Estimate segment field usages (#112760)

We have introduced a new memory estimation method in serverless, based 
on the number of segments and the fields within them. This new approach
works well overall, but it still falls short in cases where most fields
are used more than once - for example, in both doc_values and postings,
or doc_values and points. This change exposes the total usage of fields
in segments, allowing us to adjust the memory estimate for these cases.

ywangd committed:
Use a dedicated test executor in MockTransportService (#112748)

Instead of using the generic executor for delayed transport actions,
this PR adds a new executor to schedule these actions. It helps avoid
sharing executors with the node which may lead to unexpected CI failures
due to unsafe future assertion.

elasticsearchmachine committed:
Mute org.elasticsearch.repositories.blobstore.testkit.integrity.RepositoryVerifyIntegrityIT testTransportException #112779

elasticsearchmachine committed:
Mute org.elasticsearch.script.StatsSummaryTests testEqualsAndHashCode #112439

ywangd committed:
Do not throw in task enqueued by CancellableRunner (#112780)

CancellableThreads#excute can throw runtime exception including
cancellation. This does not work with AbstractThrottledTaskRunner which
expects enqueued task to _not_ throw. This PR catches any runtime
exception from CancellableThreads and hand it back to the original
runnable.

Resolves: #112779

ywangd committed:
[Test] Account for auto-repairing for shard gen file (#112778)

Since #112337, missing shard gen files are automatically reconstructed
based on the existing shard snapshot files. If the list of shard
snapshot files is complete, it means the repository is effectively not
corrupted. This PR updates the test to account for this situation.

Resolves: #112769

DaveCTurner committed:
Introduce test utils for ingest pipelines (#112733)

Replaces the somewhat-awkward API on `ClusterAdminClient` for
manipulating ingest pipelines with some test-specific utilities that are
easier to use.

Relates #107984 in that this change massively reduces the noise that
would otherwise result from removing the trappy timeouts in these APIs.

iverase committed:
Deduplicate BucketOrder when deserializing (#112707)

Deduplicate BucketOrder object by wrapping the StreamInput generated by DelayableWritable objects.

thecoop committed:
Block use of current version feature in yaml tests (#112737)

Using the current version as a condition is not compatible with Serverless, as the version doesn't change between commits. Cluster features should be used instead.

v1v committed:
ci(bump automation): bump ubi9 for ironbank (#112298)

### What

Enable updatecli
[policies](https://www.updatecli.io/blog/shareable-policies) to bump the
Ironbank versions automatically, then 
https://github.com/elastic/elasticsearch/pull/111743 won't be manually
created but when a new ubit9 version is released and available in the
Ironbank system.

Those policies can be found at
[elastic/oblt-updatecli-policies@main/updatecli/policies/](https://github.com/elastic/oblt-updatecli-policies/tree/main/updatecli/policies/?rgh-link-date=2024-08-28T16%3A30%3A04Z)
(NOTE: This is a private repository only accessible by Elastic
employees)

#### How to test this PR locally

1. `gh pr checkout 112298`
2. Install [updatecli](https://www.updatecli.io/docs/prologue/installation/)
3. Login to [ghcr.io](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#authenticating-to-the-container-registry)
4. Diff (dry-run)

```bash
$ GITHUB_TOKEN=$(gh auth token) updatecli compose diff --experimental
```

5. Create Pull Request if new changes

```bash
$ GITHUB_REPOSITORY=elastic/elasticsearch \
   GITHUB_ACTOR=v1v \
   GITHUB_TOKEN=$(gh auth token) \
updatecli compose apply --experimental
```

piergm committed:
Two empty mappings now are created equally (#107936)

* Two empty mappings reported equally in field_caps

* empty mapping now is {}

* iter

DaveCTurner committed:
Cleanup shutdown module bwc in v9 (#112793)

Removes a bunch of now-unnecessary bwc code in the shutdown module.

thecoop committed:
Update last few references in yaml tests from ROOT locale to ENGLISH (#112791)

jan-elastic committed:
Remove adaptive allocations feature flag (#112798)

davidkyle committed:
address comments

davidkyle commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into background-download-write

davidkyle committed:
fix recovery from failure

davidkyle commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into background-download-write
